### ISSUE TITLE
`WeightSetVisitor` and `WeightSizeVisitor` should calculate the size of the layer.

### ISSUE THREAD

**Issue by himanshupathak21061998** [mlpack/mlpack#2590]
#### Issue description
`WeightSetVisitor` and `WeightSizeVisitor` should calculate the size of each layer by using the parameters like `inSize` and `outSize` but currently they are not doing same. It should be different for each layer.
For Linear Layer it will be `inSize * outSize + outSize` .
Note:- While adding implementation of WeigthSize please look at constructor to get the information about the size of weight matrix of that layer.
Related to https://github.com/mlpack/mlpack/pull/2324#issuecomment-679448653

 A list of all the layers -
 
  * [x]  add.hpp
  * [ ]  add_merge.hpp
  * [x]  adaptive_max_pooling.hpp
  * [x]  adaptive_mean_pooling.hpp
  * [ ]  alpha_dropout.hpp
  * [x]  atrous_convolution.hpp
  * [ ]  base_layer.hpp
  * [x]  batch_norm.hpp
  * [ ]  bilinear_interpolation.hpp
  * [x]  concat.hpp
  * [ ]  concat_performance.hpp
  * [ ]  concatenate.hpp
  * [ ]  constant.hpp
  * [x]  convolution.hpp
  * [x]  dropconnect.hpp
  * [ ]  dropout.hpp
  * [ ]  elu.hpp
  * [x]  fast_lstm.hpp
  * [ ]  flexible_relu.hpp
  * [ ]  glimpse.hpp
  * [ ]  gru.hpp
  * [ ]  hard_tanh.hpp
  * [ ]  highway.hpp
  * [ ]  join.hpp
  * [ ]  layer.hpp
  * [ ]  layer_norm.hpp
  * [ ]  layer_traits.hpp
  * [ ]  layer_types.hpp
  * [ ]  leaky_relu.hpp
  * [x]  linear.hpp
  * [ ]  linear_no_bias.hpp
  * [ ]  log_softmax.hpp
  * [ ]  lookup.hpp
  * [x]  lstm.hpp
  * [x]  max_pooling.hpp
  * [x]  mean_pooling.hpp
  * [ ]  minibatch_discrimination.hpp
  * [x]  multihead_attention.hpp
  * [x]  multiply_constant.hpp
  * [x]  multiply_merge.hpp
  * [x]  noisylinear.hpp
  * [ ]  parametric_relu.hpp
  * [ ]  positional_encoding.hpp
  * [ ]  recurrent.hpp
  * [ ]  recurrent_attention.hpp
  * [ ]  reinforce_normal.hpp
  * [ ]  reparametrization.hpp
  * [ ]  radial_basis_function.hpp
  * [ ]  select.hpp
  * [x]  sequential.hpp
  * [x]  softmax.hpp
  * [ ]  spatial_dropout.hpp
  * [ ]  subview.hpp
  * [x]  transposed_convolution.hpp
  * [ ]  vr_class_reward.hpp
  * [ ]  c_relu.hpp
  * [x]  weight_norm.hpp
  * [ ]  hardshrink.hpp
  * [ ]  celu.hpp
  * [ ]  softshrink.hpp

**Comment by rcurtin** [comment]
For anyone interested in working on this, a couple more details:

`WeightSetVisitor` and `WeightSizeVisitor` in `src/mlpack/methods/ann/visitor/` are used to traverse the layers of a neural network and determine the size of the weights associated with them.  For instance, this function in `weight_size_visitor_impl.hpp` just uses `layer->Parameters().n_elem` to compute the size of the layer's weights:

```
template<typename T, typename P>
inline typename std::enable_if<
    HasParametersCheck<T, P&(T::*)()>::value &&
    !HasModelCheck<T>::value, size_t>::type
WeightSizeVisitor::LayerSize(T* layer, P& /* output */) const
{
  return layer->Parameters().n_elem;
}
```

But the goal is to, instead of using `layer->Parameters()`, perhaps implement another function in each layer that tells us the number of elements in the weights based on the parameters (like for `Linear<>` this can be computed with `inSize` and `outSize`), and use that inside of `WeightSizeVisitor::LayerSize()`.

Once we do that for every layer, this helps make the serialized models a lot smaller. :+1:

**Comment by UtR491** [comment]
Just to be sure, right now LayerSize() calls the entire weight matrix (using the Parameters() function) just to find out its size and the objective is to implement a function inside each layer which returns the size directly. Is this right? Can't this be done by just implementing a function inside each layer that returns `weights.n_elem` instead of calculating the size from parameters like `inSize` and `outSize`. I don't know a lot about serialization, so I don't if this will help achieving the main goal of reducing the size of serialized models. In any case, I would like to work on this if no one is already on it.

**Comment by rcurtin** [comment]
@UtR491 right, the issue is not that `LayerSize()` accesses the internal `weights` member of a layer, but instead the issue is that we'd like to be able to compute the size required to hold the weights of a layer *without* using the `weights` member.

The reason for this is that we'd ideally like to be able to avoid serializing the `weights` member.  (It will save space.  Instea dof having each layer serialize its `weights` member, we only have the enclosing `FFN` serialize, as one object, *all* the weights of the network.)

After loading a model from disk, the `FFN` class uses the `WeightSetVisitor` and `WeightSizeVisitor`.  But in order to avoid serializing the `weights` member in each layer, we would need `LayerSize()` to compute the size that `weights` *should* be---without actually using the `weights` member.

So, hopefully this helps clarify the issue!  Let me know if there's anything more I can explain. :+1:  There are a lot of layers, so probably we should make a list at the top of all the layers, so we can keep track of which ones are done and which ones aren't done.  For now, though, I think nothing is done quite yet.  If you want to take a first stab at, say, `Linear<>`, feel free! :)

**Comment by UtR491** [comment]
@rcurtin, I have made a PR. Please take a look.

**Comment by UtR491** [comment]
A list of all the layers -

- [x] add.hpp
- [ ] add_merge.hpp
- [ ] adaptive_max_pooling.hpp
- [ ] adaptive_mean_pooling.hpp
- [ ] alpha_dropout.hpp
- [x] atrous_convolution.hpp
- [ ] base_layer.hpp
- [ ] batch_norm.hpp
- [ ] bilinear_interpolation.hpp
- [ ] concat.hpp
- [ ] concat_performance.hpp
- [ ] concatenate.hpp
- [ ] constant.hpp
- [ ] convolution.hpp
- [ ] dropconnect.hpp
- [ ] dropout.hpp
- [ ] elu.hpp
- [ ] fast_lstm.hpp
- [ ] flexible_relu.hpp
- [ ] glimpse.hpp
- [ ] gru.hpp
- [ ] hard_tanh.hpp
- [ ] highway.hpp
- [ ] join.hpp
- [ ] layer.hpp
- [ ] layer_norm.hpp
- [ ] layer_traits.hpp
- [ ] layer_types.hpp
- [ ] leaky_relu.hpp
- [x] linear.hpp
- [ ] linear_no_bias.hpp
- [ ] log_softmax.hpp
- [ ] lookup.hpp
- [ ] lstm.hpp
- [ ] max_pooling.hpp
- [ ] mean_pooling.hpp
- [ ] minibatch_discrimination.hpp
- [ ] multihead_attention.hpp
- [ ] multiply_constant.hpp
- [ ] multiply_merge.hpp
- [ ] noisylinear.hpp
- [ ] parametric_relu.hpp
- [ ] positional_encoding.hpp
- [ ] recurrent.hpp
- [ ] recurrent_attention.hpp
- [ ] reinforce_normal.hpp
- [ ] reparametrization.hpp
- [ ] radial_basis_function.hpp
- [ ] select.hpp
- [ ] sequential.hpp
- [ ] softmax.hpp
- [ ] spatial_dropout.hpp
- [ ] subview.hpp
- [ ] transposed_convolution.hpp
- [ ] vr_class_reward.hpp
- [ ] c_relu.hpp
- [ ] weight_norm.hpp
- [ ] hardshrink.hpp
- [ ] celu.hpp
- [ ] softshrink.hpp

**Comment by kartikdutt18** [comment]
Taking up BatchNorm and Convolution layers. Thanks for opening the issue.

**Comment by himanshupathak21061998** [comment]
Taking up Weight_Norm and sequential.

**Comment by codeboy5** [comment]
Taking up dropout. thanks !

**Comment by RishabhGarg108** [comment]
Just to clarify, here by calculating the size of each layer, do we mean the number of parameters associated with a particular layer ? And do we also consider bias terms along with weight terms while calculating this number ?

Alse, correct me if I am wrong here, but shouldn't the parameters for the activation layers like softmax, tanh, CRELU etc. be zero ?

**Comment by rcurtin** [comment]
@RishabhGarg108 right on all accounts. :+1: So bias terms should be considered too, and for layers like softmax, the `WeightSize()` should indeed return 0.

**Comment by RishabhGarg108** [comment]
Ok @rcurtin , thanks.

Shall I take max_pooling, min_pooling and softmax layers if no one is currently working on them ?

**Comment by rcurtin** [comment]
Sure, go for it!

**Comment by RishabhGarg108** [comment]
Hey, I was looking at the recurrent.hpp file and there are no inSize and outSize parameters. But we have startModule, inputModule, feedbackModule and transferModule. All these modules are layers of type CustomLayers and I do not know how to extract the required dimensions of matrices from them. It would be nice if  someone could guide me how to work with these modules.

**Comment by rcurtin** [comment]
Hey @RishabhGarg108, sorry for the slow response here.  You are right that the `Recurrent` layer is just a wrapper around other layers, so it doesn't hold weights directly.  The visitor framework will visit the child modules correctly, so you can just return `0` in that case.  (If you're interested in how that's done, take a look at `src/mlpack/methods/ann/visitor/weight_size_visitor_impl.hpp` where it checks whether or not the layer has the `Model()` function, and see how that's used.) :+1:

**Comment by RishabhGarg108** [comment]
Thanks @rcurtin ! I will definitely look at the implementation to see how those modules are used. :+1:

**Comment by gauravghati** [comment]
Hey, Taking up GRU and adaptive max and mean pooling. 
@rcurtin, could you keep this issue open?

**Comment by ayushsingh11** [comment]
Hi all, can I take up multihead_attention, multiply_constant and multiply_merge? @rcurtin

**Comment by hello-fri-end** [comment]
Hi! Can I take up minibatch_discrimination and bilinear_interpolation? @rcurtin

**Comment by rcurtin** [comment]
If it hasn't been done yet, feel free to open a PR. :+1:

**Comment by onikolskyy** [comment]
https://github.com/mlpack/mlpack/pull/2859

Added WeightSize function for `linear_no_bias`, `rbf`, `constant`.
